posts/2024-04-20-Performance Comparison: Data.table in R and Pandas in Python-Doris Amoakohene/index.qmd

```diff
@@ -0,0 +1,372 @@
+
+ml.colors <- c(
+  "data.table::melt"="#D6604D",
+  "pandas::pd.melt" = "#BF812D"
+  )
+n.folds <- 10
+n.rows <- 100
+seconds.limit <- 10
+
+ml.reshape.atime <- atime::atime(
+  N=as.integer(10^seq(2, 15, by=0.2)),
+  setup={
+    df <- data.frame(
+      id = rep(1:N, each = 2),
+      category = rep(c("A", "B"), N),
+      value = rnorm(2 * N)
+      )
+    },
+  seconds.limit= seconds.limit,
+  
+  "data.table::melt" = {
+    data.table::melt(data.table(df), id.vars = c("id",  "category"),variable.names="variable", value.name = "value")
+  },
+  "pandas::pd.melt" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
+    pd$melt(py_df, id_vars = c("id", "category"), value_name = "score")  
+  }
+  )
+
+```
+
+```{r,warning=FALSE,message=FALSE}
+ml.reshape.refs <- atime::references_best(ml.reshape.atime)
+plot(ml.reshape.refs)
+```
+Also in this graph we see that data.table::melt performs for complexity notations to pandas::melt
+
+
+```{r,warning=FALSE,message=FALSE}
+ml.reshape.pred <- predict(ml.reshape.refs)
+plot(ml.reshape.pred)
+```
+
+
+```{r,warning=FALSE,message=FALSE}
+ml.wide2long.pd <- plot(ml.reshape.pred)+
+  theme(text=element_text(size=15))+
+  ggtitle(sprintf("Reshaping from wide to long panda & data.table \nover real numbers, N times", n.folds))+
+  scale_x_log10("N = number of Mean,SD,Length to compute")+
+  scale_y_log10("Computation time (seconds)
+median line, min/max band
+over 10 timings")+
+  facet_null()+
+  scale_fill_manual(values=ml.colors)+
+  scale_color_manual(values=ml.colors)
+```
+
+```{r, warning=FALSE, message=FALSE}
+plot(ml.wide2long.pd)
+```
+
+When converting data from wide to long format, as shown in the above plot, data.table's melt() function efficiently gathers multiple columns into key-value pairs, allowing for easy transformation of wide data into a longer, more structured format. 
+
+
+
+## B. long to wide reshape
+
+In data.table, the data.table::dcast() function is often used to convert data from a long format to a wide format, and in pandas, the pandas::pivot_table() function is used to convert data from a long format to a wide format.
+
+## Comparison code
+
+## data.table::dcast() is faster
+
+```{r,warning=FALSE,message=FALSE,results='hide'}
+
+ml.colors <- c(
+  "data.table::dcast" = "#D6604D",
+  "pandas::pivot_table" = "#BF812D"
+)
+
+n.folds <- 10
+n.rows <- 100
+seconds.limit <- 1
+
+ml.long2wide.atime <- atime::atime(
+  N=as.integer(10^seq(2, 7, by=0.2)),
+  setup={
+    df <- data.frame(
+      id = rep(1:N, each = 2),
+      category = rep(c("A", "B"), N),
+      value = rnorm(2 * N)
+      )
+    },
+  seconds.limit= seconds.limit,
+  "data.table::dcast" = {
+    data.table::dcast(data.table(df), id ~ category, value.var = "value")
+  },
+  "pandas::pivot_table" = {
+    py_df <- reticulate::r_to_py(df)
+    pd <- import("pandas")
+    pd$pivot_table(py_df, values = "value", index = "id", columns = "category")
+  }
+  )
+
```

Comment:
For code blocks too consistent indentation would be nice

like leave no extra line(s) at the end ideally (some blocks I see that, some I don't)